[Bug/错误]: signature_delta attached to wrong block in Anthropic response

### Pre-submission Checklist / 提交前检查

- [x] I have searched the [existing issues](https://github.com/looplj/axonhub/issues) and this bug has not been reported / 我已搜索现有 issues，此 Bug 尚未被报告
- [x] I have read the [documentation](https://github.com/looplj/axonhub#readme) / 我已阅读文档
- [x] I am using the latest version of AxonHub / 我正在使用 AxonHub 的最新版本

### Bug Description / Bug 描述

通过 AxonHub 使用 Claude Code，发现经常一个 prompt 发起两个请求，一个流式，一个非流式。

`claude --debug` 的日志中可以看到 streaming error 的错误：
```
2026-03-19T11:34:48.362Z [DEBUG] Stream started - received first chunk
2026-03-19T11:34:49.881Z [ERROR] Error streaming, falling back to non-streaming mode: Content block is not a thinking block
```

最近的 Claude Code 更新，支持了显示 streaming 的内容，经常会发现常常流式输出完后，突然整个文本块变了。
即虽然流式产生了错误，但客户端仍在继续显示，并在后续同步请求完成后替换了这个文本快。

### Steps to Reproduce / 复现步骤

1. 在 Claude Code 里通过 AxonHub 使用 Opus，可以使用 `claude --debug` 打开调试日志
2. `/effort high`
3. prompt 「请帮我看看这个目录下有什么」
4. 稍等片刻，在 AxonHub 里看到至少两个请求，一个是流式的，一个是非流式的。此时，`claude --debug` 的日志文件中也可见到错误。
```
2026-03-19T11:34:48.362Z [DEBUG] Stream started - received first chunk
2026-03-19T11:34:49.881Z [ERROR] Error streaming, falling back to non-streaming mode: Content block is not a thinking block
```

### Expected Behavior / 期望行为

有个空的 thinking block, signature 在这个 thinking block，然后才是 tool_use。

```
event: message_start
data: {"type":"message_start","message":{"model":"claude-opus-4-6","id":"msg_111111111111111111111111","type":"message","role":"assistant","content":[],"stop_reason":null,"stop_sequence":null,"usage":{"input_tokens":3,"cache_creation_input_tokens":0,"cache_read_input_tokens":24857,"cache_creation":{"ephemeral_5m_input_tokens":0,"ephemeral_1h_input_tokens":0},"output_tokens":11,"service_tier":"standard","inference_geo":"not_available"}}             }

event: content_block_start
data: {"type":"content_block_start","index":0,"content_block":{"type":"thinking","thinking":"","signature":""}     }

event: ping
data: {"type": "ping"}

event: content_block_delta
data: {"type":"content_block_delta","index":0,"delta":{"type":"signature_delta","signature":"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"}               }

event: content_block_stop
data: {"type":"content_block_stop","index":0            }

event: content_block_start
data: {"type":"content_block_start","index":1,"content_block":{"type":"tool_use","id":"toolu_AAAAAAAAAAAAAAAAAAAAAAAA","name":"Bash","input":{},"caller":{"type":"direct"}}     }

event: content_block_delta
data: {"type":"content_block_delta","index":1,"delta":{"type":"input_json_delta","partial_json":""} }

--- 省略若干 content_block_delta  ---

event: content_block_delta
data: {"type":"content_block_delta","index":1,"delta":{"type":"input_json_delta","partial_json":"tory\"}"}         }

event: content_block_stop
data: {"type":"content_block_stop","index":1        }
```

### Actual Behavior / 实际行为

空的 thinking block 丢了，signature 跑到后面的 tool_use block 去了。

```
event:message_start
data: {"type":"message_start","message":{"id":"msg_AAAAAAAAAAAAAAAAAAAAAAAA","type":"message","role":"assistant","content":[],"model":"claude-opus-4-6","usage":{"input_tokens":3,"output_tokens":11,"cache_creation_input_tokens":14929,"cache_read_input_tokens":10107,"cache_creation":{"ephemeral_5m_input_tokens":14929,"ephemeral_1h_input_tokens":0}}}}

event:content_block_start
data: {"type":"content_block_start","index":0,"content_block":{"type":"tool_use","id":"toolu_BBBBBBBBBBBBBBBBBBBBBBBB","name":"Bash","input":{}}}

event:content_block_delta
data: {"type":"content_block_delta","index":0,"delta":{"type":"input_json_delta","partial_json":""}}

--- 省略若干 content_block_delta  ---

event:content_block_delta
data: {"type":"content_block_delta","index":0,"delta":{"type":"input_json_delta","partial_json":"ory\"}"}}

event:content_block_delta
data: {"type":"content_block_delta","index":0,"delta":{"type":"signature_delta","signature":"QVhOMTAxAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"}}
```

### Operating System / 操作系统

Linux (Other) / Linux（其他）

### AxonHub Version / AxonHub 版本

v0.9.18

### Usage Scenario / 使用场景

Claude Code

### API Format / API 格式

Anthropic - Messages (/v1/messages) / Anthropic Messages

### Channel Type / 渠道类型

Anthropic / Anthropic

### Logs / 日志

[redacted_axonhub_response.txt](https://github.com/user-attachments/files/26114604/redacted_axonhub_response.txt)
[redacted_upstream_response.txt](https://github.com/user-attachments/files/26114605/redacted_upstream_response.txt)

### Additional Context / 其他补充信息

<details><summary>Claude Opus 4.6 对 @anthropic-ai/claude-agent-sdk 的分析</summary>

根据代码分析，错误信息 `"Content block is not a thinking block"` 来自 streaming 事件处理中的 `content_block_delta` 分支。具体在这段逻辑中：

```javascript
case "content_block_delta": {
    let j8 = $6[J8.index]; // 根据 index 获取之前 content_block_start 创建的 block
    
    switch (J8.delta.type) {
        // ...
        case "signature_delta":
            if (j8.type !== "thinking")
                throw Error("Content block is not a thinking block");
            j8.signature = J8.delta.signature;
            break;
        case "thinking_delta":
            if (j8.type !== "thinking")
                throw Error("Content block is not a thinking block");
            j8.thinking += J8.delta.thinking;
            break;
    }
}
```

## 触发原因

API 在 streaming 过程中发送了一个 `thinking_delta` 或 `signature_delta` 事件，但该事件对应的 `content_block` index 指向的块**不是 `thinking` 类型**。

也就是说：
1. `content_block_start` 事件为某个 index 创建了一个非 thinking 类型的块（比如 `text` 或 `tool_use`）
2. 随后收到的 `thinking_delta` / `signature_delta` 却引用了同一个 index

这是一个 **API 服务端 streaming 的 bug**——content block 的 index 或类型出现了不一致。客户端的类型校验发现了这个不匹配，抛出异常后走到了 catch 分支：

```javascript
} catch (t) {
    // ...
    k(`Error streaming, falling back to non-streaming mode: ${s6(t)}`, {level: "error"});
    // --> 这就是你看到的那条日志
    // 然后发起 non-streaming fallback 请求 (CNq)
}
```

## 总结

| 项目 | 详情 |
|------|------|
| **错误来源** | 客户端 streaming 类型校验 |
| **直接原因** | 收到 `thinking_delta` 或 `signature_delta`，但对应 index 的 content block 不是 `thinking` 类型 |
| **根因** | API 服务端 streaming 返回了 index/类型不匹配的事件（服务端 bug 或瞬态错误） |
| **后续行为** | 客户端捕获异常，回退到 non-streaming 模式重新发送请求 |

这不是你这边能修的问题，是 Anthropic API streaming 端的一个偶发异常。fallback 机制正是为了应对这类情况——streaming 失败后用 non-streaming 兜底，保证请求最终能成功。
</details>

--- a/llm/transformer/anthropic/inbound_convert.go
+++ b/llm/transformer/anthropic/inbound_convert.go
@@ -304,9 +304,17 @@ func convertToLLMRequest(anthropicReq *MessageRequest) (*llm.Request, error) {
 		case "enabled":
 			chatReq.ReasoningEffort = thinkingBudgetToReasoningEffort(anthropicReq.Thinking.BudgetTokens)
 			chatReq.ReasoningBudget = lo.ToPtr(anthropicReq.Thinking.BudgetTokens)
+
+			if anthropicReq.Thinking.Display != "" {
+				chatReq.TransformerMetadata[TransformerMetadataKeyThinkingDisplay] = anthropicReq.Thinking.Display
+			}
 		case "adaptive":
 			// Adaptive thinking doesn't require a budget; preserve the type marker via TransformerMetadata.
 			chatReq.TransformerMetadata[TransformerMetadataKeyThinkingType] = "adaptive"
+
+			if anthropicReq.Thinking.Display != "" {
+				chatReq.TransformerMetadata[TransformerMetadataKeyThinkingDisplay] = anthropicReq.Thinking.Display
+			}
 		}
 	}
 
@@ -375,9 +383,13 @@ func convertToAnthropicResponse(chatResp *llm.Response) *Message {
 
 			// Handle reasoning content (thinking) first if present
 			if (message.ReasoningContent != nil && *message.ReasoningContent != "") || (message.ReasoningSignature != nil && *message.ReasoningSignature != "") {
+				thinkingContent := message.ReasoningContent
+				if thinkingContent == nil {
+					thinkingContent = lo.ToPtr("")
+				}
 				thinkingBlock := MessageContentBlock{
 					Type:     "thinking",
-					Thinking: message.ReasoningContent,
+					Thinking: thinkingContent,
 				}
 				if message.ReasoningSignature != nil {
 					thinkingBlock.Signature = message.ReasoningSignature

--- a/llm/transformer/anthropic/inbound_stream.go
+++ b/llm/transformer/anthropic/inbound_stream.go
@@ -53,23 +53,82 @@ type anthropicInboundStream struct {
 	pendingSignature *string
 }
 
-// flushPendingSignature emits a buffered signature_delta event.
-// Call this before closing a thinking block so the signature appears after thinking content.
-func (s *anthropicInboundStream) flushPendingSignature() error {
-	if s.pendingSignature == nil {
+// closeThinkingBlock ensures any open or implied thinking block is properly
+// closed. It handles three scenarios:
+//  1. pendingSignature exists but no thinking block was started — creates a
+//     synthetic empty thinking block (start + signature_delta + stop).
+//  2. A thinking block is open — flushes any pending signature as
+//     signature_delta, then emits content_block_stop.
+//  3. Neither — no-op.
+func (s *anthropicInboundStream) closeThinkingBlock() error {
+	if s.pendingSignature != nil && !s.hasThinkingContentStarted {
+		sig := s.pendingSignature
+		s.pendingSignature = nil
+
+		if err := s.enqueEvent(&StreamEvent{
+			Type:  "content_block_start",
+			Index: &s.contentIndex,
+			ContentBlock: &MessageContentBlock{
+				Type:     "thinking",
+				Thinking: lo.ToPtr(""),
+			},
+		}); err != nil {
+			return fmt.Errorf("failed to enqueue thinking content_block_start for pending signature: %w", err)
+		}
+
+		if err := s.enqueEvent(&StreamEvent{
+			Type:  "content_block_delta",
+			Index: &s.contentIndex,
+			Delta: &StreamDelta{
+				Type:      lo.ToPtr("signature_delta"),
+				Signature: sig,
+			},
+		}); err != nil {
+			return fmt.Errorf("failed to enqueue signature_delta for pending signature: %w", err)
+		}
+
+		if err := s.enqueEvent(&StreamEvent{
+			Type:  "content_block_stop",
+			Index: &s.contentIndex,
+		}); err != nil {
+			return fmt.Errorf("failed to enqueue content_block_stop for pending signature: %w", err)
+		}
+
+		s.contentIndex += 1
+
 		return nil
 	}
-	sig := s.pendingSignature
-	s.pendingSignature = nil
 
-	return s.enqueEvent(&StreamEvent{
-		Type:  "content_block_delta",
-		Index: &s.contentIndex,
-		Delta: &StreamDelta{
-			Type:      lo.ToPtr("signature_delta"),
-			Signature: sig,
-		},
-	})
+	if s.hasThinkingContentStarted {
+		s.hasThinkingContentStarted = false
+
+		if s.pendingSignature != nil {
+			sig := s.pendingSignature
+			s.pendingSignature = nil
+
+			if err := s.enqueEvent(&StreamEvent{
+				Type:  "content_block_delta",
+				Index: &s.contentIndex,
+				Delta: &StreamDelta{
+					Type:      lo.ToPtr("signature_delta"),
+					Signature: sig,
+				},
+			}); err != nil {
+				return fmt.Errorf("failed to enqueue signature_delta event: %w", err)
+			}
+		}
+
+		if err := s.enqueEvent(&StreamEvent{
+			Type:  "content_block_stop",
+			Index: &s.contentIndex,
+		}); err != nil {
+			return fmt.Errorf("failed to enqueue content_block_stop event: %w", err)
+		}
+
+		s.contentIndex += 1
+	}
+
+	return nil
 }
 
 func (s *anthropicInboundStream) enqueEvent(ev *StreamEvent) error {
@@ -244,27 +303,9 @@ func (s *anthropicInboundStream) Next() bool {
 
 		// Handle redacted reasoning content (redacted_thinking)
 		if choice.Delta != nil && choice.Delta.RedactedReasoningContent != nil && *choice.Delta.RedactedReasoningContent != "" {
-			// If the thinking content has started before the redacted thinking content, we need to stop it
-			if s.hasThinkingContentStarted {
-				s.hasThinkingContentStarted = false
-
-				if err := s.flushPendingSignature(); err != nil {
-					s.err = fmt.Errorf("failed to flush pending signature: %w", err)
-					return false
-				}
-
-				stopEvent := StreamEvent{
-					Type:  "content_block_stop",
-					Index: &s.contentIndex,
-				}
-
-				err := s.enqueEvent(&stopEvent)
-				if err != nil {
-					s.err = fmt.Errorf("failed to enqueue content_block_stop event: %w", err)
-					return false
-				}
-
-				s.contentIndex += 1
+			if err := s.closeThinkingBlock(); err != nil {
+				s.err = fmt.Errorf("failed to close thinking block: %w", err)
+				return false
 			}
 
 			// If the tool content has started before the redacted thinking content, we need to stop it
@@ -333,27 +374,9 @@ func (s *anthropicInboundStream) Next() bool {
 
 		// Handle content delta
 		if choice.Delta != nil && choice.Delta.Content.Content != nil && *choice.Delta.Content.Content != "" {
-			// If the thinking content has started before the text content, we need to stop it
-			if s.hasThinkingContentStarted {
-				s.hasThinkingContentStarted = false
-
-				if err := s.flushPendingSignature(); err != nil {
-					s.err = fmt.Errorf("failed to flush pending signature: %w", err)
-					return false
-				}
-
-				stopEvent := StreamEvent{
-					Type:  "content_block_stop",
-					Index: &s.contentIndex,
-				}
-
-				err := s.enqueEvent(&stopEvent)
-				if err != nil {
-					s.err = fmt.Errorf("failed to enqueue content_block_stop event: %w", err)
-					return false
-				}
-
-				s.contentIndex += 1
+			if err := s.closeThinkingBlock(); err != nil {
+				s.err = fmt.Errorf("failed to close thinking block: %w", err)
+				return false
 			}
 
 			// If the tool content has started before the content block, we need to stop it
@@ -413,28 +436,9 @@ func (s *anthropicInboundStream) Next() bool {
 
 		// Handle tool calls
 		if choice.Delta != nil && len(choice.Delta.ToolCalls) > 0 {
-			// Support tool call when the thinking.
-			// If the thinking content has started before the text content, we need to stop it
-			if s.hasThinkingContentStarted {
-				s.hasThinkingContentStarted = false
-
-				if err := s.flushPendingSignature(); err != nil {
-					s.err = fmt.Errorf("failed to flush pending signature: %w", err)
-					return false
-				}
-
-				stopEvent := StreamEvent{
-					Type:  "content_block_stop",
-					Index: &s.contentIndex,
-				}
-
-				err := s.enqueEvent(&stopEvent)
-				if err != nil {
-					s.err = fmt.Errorf("failed to enqueue content_block_stop event: %w", err)
-					return false
-				}
-
-				s.contentIndex += 1
+			if err := s.closeThinkingBlock(); err != nil {
+				s.err = fmt.Errorf("failed to close thinking block: %w", err)
+				return false
 			}
 
 			// If the text content has started before the tool content, we need to stop it
@@ -553,8 +557,8 @@ func (s *anthropicInboundStream) Next() bool {
 		if choice.FinishReason != nil && !s.hasFinished {
 			s.hasFinished = true
 
-			if err := s.flushPendingSignature(); err != nil {
-				s.err = fmt.Errorf("failed to flush pending signature: %w", err)
+			if err := s.closeThinkingBlock(); err != nil {
+				s.err = fmt.Errorf("failed to close thinking block: %w", err)
 				return false
 			}
 

--- a/llm/transformer/anthropic/inbound_stream_pending_signature_test.go
+++ b/llm/transformer/anthropic/inbound_stream_pending_signature_test.go
@@ -64,6 +64,23 @@ func withTextContent(text string) func(*llm.Response) {
 	}
 }
 
+func withToolCall(index int, id, name, args string) func(*llm.Response) {
+	return func(r *llm.Response) {
+		if r.Choices[0].Delta == nil {
+			r.Choices[0].Delta = &llm.Message{Role: "assistant"}
+		}
+		r.Choices[0].Delta.ToolCalls = append(r.Choices[0].Delta.ToolCalls, llm.ToolCall{
+			Index: index,
+			ID:    id,
+			Type:  "function",
+			Function: llm.FunctionCall{
+				Name:      name,
+				Arguments: args,
+			},
+		})
+	}
+}
+
 func withFinishReason(reason string) func(*llm.Response) {
 	return func(r *llm.Response) {
 		r.Choices[0].FinishReason = lo.ToPtr(reason)
@@ -294,4 +311,161 @@ func TestPendingSignature_NoSignature(t *testing.T) {
 			require.NotEqual(t, "signature_delta", *ev.Delta.Type, "signature_delta should not appear without signature")
 		}
 	}
+}
+
+// TestPendingSignature_SignatureWithoutThinking_ToolUse tests the defensive path:
+// signature arrives with no ReasoningContent at all, then tool_use directly.
+// The flushPendingSignatureBlock() creates a synthetic empty thinking block.
+func TestPendingSignature_SignatureWithoutThinking_ToolUse(t *testing.T) {
+	const (
+		id    = "msg_test_006"
+		model = "test-model"
+		sig   = "orphan_sig_tool"
+	)
+
+	responses := []*llm.Response{
+		buildChunk(id, model, withUsage(10, 1)),
+		// Signature without any thinking content
+		buildChunk(id, model, withReasoningSignature(sig)),
+		// Tool use directly
+		buildChunk(id, model, withToolCall(0, "toolu_01", "Bash", `{"command":"ls"}`)),
+		buildChunk(id, model, withFinishReason("tool_calls")),
+		buildChunk(id, model, withUsage(10, 20)),
+	}
+
+	events := collectStreamEvents(t, responses)
+
+	// Expected: synthetic thinking block created for the signature
+	// 0: message_start
+	// 1: content_block_start (thinking)
+	// 2: content_block_delta (signature_delta)
+	// 3: content_block_stop (index 0)
+	// 4: content_block_start (tool_use, index 1)
+	// 5: content_block_delta (input_json_delta)
+	// 6: content_block_stop (index 1)
+	// 7: message_delta
+	// 8: message_stop
+
+	require.Len(t, events, 9)
+
+	// Synthetic thinking block
+	require.Equal(t, "content_block_start", events[1].Type)
+	require.Equal(t, "thinking", events[1].ContentBlock.Type)
+	require.Equal(t, int64(0), *events[1].Index)
+
+	// Signature on the thinking block
+	require.Equal(t, "content_block_delta", events[2].Type)
+	require.Equal(t, "signature_delta", *events[2].Delta.Type)
+	require.Equal(t, sig, *events[2].Delta.Signature)
+	require.Equal(t, int64(0), *events[2].Index)
+
+	require.Equal(t, "content_block_stop", events[3].Type)
+	require.Equal(t, int64(0), *events[3].Index)
+
+	// Tool use at index 1
+	require.Equal(t, "content_block_start", events[4].Type)
+	require.Equal(t, "tool_use", events[4].ContentBlock.Type)
+	require.Equal(t, int64(1), *events[4].Index)
+}
+
+// TestPendingSignature_SignatureWithoutThinking_Text tests the defensive path:
+// signature arrives with no ReasoningContent, then text directly.
+func TestPendingSignature_SignatureWithoutThinking_Text(t *testing.T) {
+	const (
+		id    = "msg_test_007"
+		model = "test-model"
+		sig   = "orphan_sig_text"
+	)
+
+	responses := []*llm.Response{
+		buildChunk(id, model, withUsage(10, 1)),
+		// Signature without any thinking content
+		buildChunk(id, model, withReasoningSignature(sig)),
+		// Text directly
+		buildChunk(id, model, withTextContent("Hello")),
+		buildChunk(id, model, withFinishReason("stop")),
+		buildChunk(id, model, withUsage(10, 10)),
+	}
+
+	events := collectStreamEvents(t, responses)
+
+	// Expected: synthetic thinking block, then text
+	// 0: message_start
+	// 1: content_block_start (thinking)
+	// 2: content_block_delta (signature_delta)
+	// 3: content_block_stop (index 0)
+	// 4: content_block_start (text, index 1)
+	// 5: content_block_delta (text_delta "Hello")
+	// 6: content_block_stop (index 1)
+	// 7: message_delta
+	// 8: message_stop
+
+	require.Len(t, events, 9)
+
+	// Synthetic thinking block
+	require.Equal(t, "content_block_start", events[1].Type)
+	require.Equal(t, "thinking", events[1].ContentBlock.Type)
+	require.Equal(t, int64(0), *events[1].Index)
+
+	// Signature on the thinking block
+	require.Equal(t, "content_block_delta", events[2].Type)
+	require.Equal(t, "signature_delta", *events[2].Delta.Type)
+	require.Equal(t, sig, *events[2].Delta.Signature)
+	require.Equal(t, int64(0), *events[2].Index)
+
+	require.Equal(t, "content_block_stop", events[3].Type)
+
+	// Text at index 1
+	require.Equal(t, "content_block_start", events[4].Type)
+	require.Equal(t, "text", events[4].ContentBlock.Type)
+	require.Equal(t, int64(1), *events[4].Index)
+}
+
+// TestPendingSignature_SignatureWithoutThinking_FinishOnly tests the defensive path:
+// signature arrives with no ReasoningContent, then finish directly (no text/tool).
+// The flushPendingSignatureBlock() at finish_reason creates a synthetic thinking block.
+func TestPendingSignature_SignatureWithoutThinking_FinishOnly(t *testing.T) {
+	const (
+		id    = "msg_test_008"
+		model = "test-model"
+		sig   = "orphan_sig_finish"
+	)
+
+	responses := []*llm.Response{
+		buildChunk(id, model, withUsage(10, 1)),
+		// Signature without any thinking content
+		buildChunk(id, model, withReasoningSignature(sig)),
+		// Finish directly
+		buildChunk(id, model, withFinishReason("stop")),
+		buildChunk(id, model, withUsage(10, 5)),
+	}
+
+	events := collectStreamEvents(t, responses)
+
+	// Expected: synthetic thinking block created at finish
+	// 0: message_start
+	// 1: content_block_start (thinking)
+	// 2: content_block_delta (signature_delta)
+	// 3: content_block_stop (index 0)
+	// 4: content_block_stop (index 0) — from finish_reason (deduplicated)
+	// But due to dedup logic, the second content_block_stop is skipped
+	// So: 4: message_delta, 5: message_stop
+
+	require.Len(t, events, 6)
+
+	// Synthetic thinking block
+	require.Equal(t, "content_block_start", events[1].Type)
+	require.Equal(t, "thinking", events[1].ContentBlock.Type)
+	require.Equal(t, int64(0), *events[1].Index)
+
+	// Signature on the thinking block
+	require.Equal(t, "content_block_delta", events[2].Type)
+	require.Equal(t, "signature_delta", *events[2].Delta.Type)
+	require.Equal(t, sig, *events[2].Delta.Signature)
+	require.Equal(t, int64(0), *events[2].Index)
+
+	require.Equal(t, "content_block_stop", events[3].Type)
+
+	require.Equal(t, "message_delta", events[4].Type)
+	require.Equal(t, "message_stop", events[5].Type)
 }

--- a/llm/transformer/anthropic/integration_test.go
+++ b/llm/transformer/anthropic/integration_test.go
@@ -20,10 +20,11 @@ func TestAnthropicTransformers_Integration(t *testing.T) {
 	outboundTransformer, _ := NewOutboundTransformer("https://api.anthropic.com", "test-api-key")
 
 	tests := []struct {
-		name                 string
-		anthropicRequestJSON string
-		expectedModel        string
-		expectedMaxTokens    int64
+		name                    string
+		anthropicRequestJSON    string
+		expectedModel           string
+		expectedMaxTokens       int64
+		expectedThinkingDisplay string
 	}{
 		{
 			name: "simple text message",
@@ -85,6 +86,88 @@ func TestAnthropicTransformers_Integration(t *testing.T) {
 			expectedModel:     "claude-3-sonnet-20240229",
 			expectedMaxTokens: 1024,
 		},
+		{
+			name: "thinking with display summarized",
+			anthropicRequestJSON: `{
+				"model": "claude-sonnet-4-20250514",
+				"max_tokens": 8096,
+				"thinking": {
+					"type": "enabled",
+					"budget_tokens": 5000,
+					"display": "summarized"
+				},
+				"messages": [
+					{
+						"role": "user",
+						"content": "Hello"
+					}
+				]
+			}`,
+			expectedModel:           "claude-sonnet-4-20250514",
+			expectedMaxTokens:       8096,
+			expectedThinkingDisplay: "summarized",
+		},
+		{
+			name: "thinking with display omitted",
+			anthropicRequestJSON: `{
+				"model": "claude-sonnet-4-20250514",
+				"max_tokens": 4096,
+				"thinking": {
+					"type": "enabled",
+					"budget_tokens": 10000,
+					"display": "omitted"
+				},
+				"messages": [
+					{
+						"role": "user",
+						"content": "Hello"
+					}
+				]
+			}`,
+			expectedModel:           "claude-sonnet-4-20250514",
+			expectedMaxTokens:       4096,
+			expectedThinkingDisplay: "omitted",
+		},
+		{
+			name: "adaptive thinking with display summarized",
+			anthropicRequestJSON: `{
+				"model": "claude-sonnet-4-20250514",
+				"max_tokens": 4096,
+				"thinking": {
+					"type": "adaptive",
+					"display": "summarized"
+				},
+				"messages": [
+					{
+						"role": "user",
+						"content": "Hello"
+					}
+				]
+			}`,
+			expectedModel:           "claude-sonnet-4-20250514",
+			expectedMaxTokens:       4096,
+			expectedThinkingDisplay: "summarized",
+		},
+		{
+			name: "disabled thinking ignores display",
+			anthropicRequestJSON: `{
+				"model": "claude-sonnet-4-20250514",
+				"max_tokens": 4096,
+				"thinking": {
+					"type": "disabled",
+					"display": "summarized"
+				},
+				"messages": [
+					{
+						"role": "user",
+						"content": "Hello"
+					}
+				]
+			}`,
+			expectedModel:           "claude-sonnet-4-20250514",
+			expectedMaxTokens:       4096,
+			expectedThinkingDisplay: "",
+		},
 	}
 
 	for _, tt := range tests {
@@ -124,6 +207,14 @@ func TestAnthropicTransformers_Integration(t *testing.T) {
 			require.NoError(t, err)
 			require.Equal(t, tt.expectedModel, anthropicReq.Model)
 			require.Equal(t, tt.expectedMaxTokens, anthropicReq.MaxTokens)
+
+			// Verify thinking display round-trip
+			if tt.expectedThinkingDisplay != "" {
+				require.NotNil(t, anthropicReq.Thinking)
+				require.Equal(t, tt.expectedThinkingDisplay, anthropicReq.Thinking.Display)
+			} else if anthropicReq.Thinking != nil {
+				require.Empty(t, anthropicReq.Thinking.Display)
+			}
 
 			// Step 3: Simulate Anthropic response and transform back
 			anthropicResponse := &Message{

--- a/llm/transformer/anthropic/model.go
+++ b/llm/transformer/anthropic/model.go
@@ -149,9 +149,14 @@ const TransformerMetadataKeyThinkingType = "thinking_type"
 // TransformerMetadataKeyOutputConfigEffort is the key for storing output config effort in TransformerMetadata.
 const TransformerMetadataKeyOutputConfigEffort = "output_config_effort"
 
+// TransformerMetadataKeyThinkingDisplay is the key for storing thinking display in TransformerMetadata.
+const TransformerMetadataKeyThinkingDisplay = "thinking_display"
+
 type Thinking struct {
 	Type         string `json:"type"          validate:"required,oneof=enabled disabled adaptive"`
 	BudgetTokens int64  `json:"budget_tokens,omitempty" validate:"required_if=Type enabled"`
+	// Display is an optional display name for the thinking, enum: summarized, omitted.
+	Display string `json:"display,omitempty"`
 }
 
 // OutputConfig represents Anthropic output configuration.

--- a/llm/transformer/anthropic/outbound_convert.go
+++ b/llm/transformer/anthropic/outbound_convert.go
@@ -83,6 +83,13 @@ func buildBaseRequest(chatReq *llm.Request, config *Config) *MessageRequest {
 		req.Thinking = buildThinking(chatReq, config)
 	}
 
+	// Restore thinking display from TransformerMetadata
+	if req.Thinking != nil && chatReq.TransformerMetadata != nil {
+		if display, ok := chatReq.TransformerMetadata[TransformerMetadataKeyThinkingDisplay].(string); ok && display != "" {
+			req.Thinking.Display = display
+		}
+	}
+
 	// Restore output_config from TransformerMetadata
 	if chatReq.TransformerMetadata != nil {
 		if effort, ok := chatReq.TransformerMetadata[TransformerMetadataKeyOutputConfigEffort].(string); ok && effort != "" {

